okhttp-tests/src/test/java/com/squareup/okhttp/internal/allocations/ConnectionTest.java

```diff
@@ -0,0 +1,207 @@
+  ConnectionPool connectionPool = new ConnectionPool(1, 1000L);
+
+  @Test public void reserveCreateCompleteRelease() throws Exception {
+    Connection connection = new Connection(connectionPool);
+    connection.setAllocationLimit(1);
+
+    Connection.StreamAllocation a = connection.reserve("a");
+    Connection.Stream a1 = a.newStream("a1");
+    assertNotNull(a1);
+    a.streamComplete(a1);
+    assertEquals(1, connection.size()); // Still allocated.
+
+    connection.release(a);
+    assertEquals(0, connection.size());
+  }
+
+  @Test public void reserveCreateReleaseComplete() throws Exception {
+    Connection connection = new Connection(connectionPool);
+    connection.setAllocationLimit(1);
+
+    Connection.StreamAllocation a = connection.reserve("a");
+    Connection.Stream a1 = a.newStream("a1");
+    assertNotNull(a1);
+    connection.release(a);
+    assertEquals(1, connection.size()); // Still allocated.
+
+    a.streamComplete(a1);
+    assertEquals(0, connection.size());
+  }
+
+  @Test public void reuseAllocation() throws Exception {
+    Connection connection = new Connection(connectionPool);
+    connection.setAllocationLimit(1);
+
+    Connection.StreamAllocation a = connection.reserve("a");
+    Connection.Stream a1 = a.newStream("a1");
+    assertNotNull(a1);
+    a.streamComplete(a1);
+    assertEquals(1, connection.size());
+
+    Connection.Stream a2 = a.newStream("a2");
+    assertNotNull(a2);
+    a.streamComplete(a2);
+    assertEquals(1, connection.size());
+
+    connection.release(a);
+    assertEquals(0, connection.size());
+  }
+
+  @Test public void cannotReuseAllocationAfterRelease() throws Exception {
+    Connection connection = new Connection(connectionPool);
+    connection.setAllocationLimit(1);
+
+    Connection.StreamAllocation a = connection.reserve("a");
+    Connection.Stream a1 = a.newStream("a1");
+    a.streamComplete(a1);
+    connection.release(a);
+
+    try {
+      a.newStream("a2");
+      fail();
+    } catch (IllegalStateException expected) {
+    }
+  }
+
+  @Test public void createReturnsNullAfterNoNewStreams() throws Exception {
+    Connection connection = new Connection(connectionPool);
+    connection.setAllocationLimit(1);
+
+    Connection.StreamAllocation a = connection.reserve("a");
+    Connection.Stream a1 = a.newStream("a1");
+    assertNotNull(a1);
+    a.streamComplete(a1);
+    assertEquals(1, connection.size());
+
+    connection.noNewStreams();
+    assertNull(a.newStream("a2"));
+    assertEquals(1, connection.size());
+
+    connection.release(a);
+    assertEquals(0, connection.size());
+  }
+
+  @Test public void reserveReturnsNullAfterNoNewStreams() throws Exception {
+    Connection connection = new Connection(connectionPool);
+    connection.setAllocationLimit(1);
+
+    Connection.StreamAllocation a = connection.reserve("a");
+    Connection.Stream a1 = a.newStream("a1");
+
+    // Setting the limit to 0 prevents new reservations.
```

Comment:
Is this comment slightly mismatched? This sets the limit as a side-effect with different semantics and is different than actually setting the limit with the `setAllocationCount` API.